spmd/__init__.py

```diff
@@ -1,103 +1,86 @@
-
-    # distribute the tensor according to DTensorSpec
-    for idx, placement in enumerate(placements):
-        if placement.is_shard():
-            placement = cast(Shard, placement)
-            shard_dim = placement.dim
-            assert (
-                shard_dim <= tensor.ndim
-            ), f"Sharding dim {shard_dim} greater than tensor ndim {tensor.ndim}"
-            # TODO: handle uneven shard sizes
-            num_chunks = device_mesh.size(dim=idx)
-            assert tensor.size(shard_dim) % num_chunks == 0, (
-                f"Only support chunk sharding evenly now, but tensor got "
-                f"dimension {shard_dim} of size {tensor.size(shard_dim)}, "
-                f"which does not divide number of shards {num_chunks}."
-            )
-            chunk_size = tensor.size(shard_dim) // num_chunks
-            tensor_list = list(tensor.chunk(num_chunks, dim=shard_dim))
-            scatter_shape = list(tensor.size())
-            scatter_shape[shard_dim] = chunk_size
-            local_tensor = device_mesh.scatter(tensor_list, mesh_dim=idx)
-            # scatter call could not return a tensor with correct requires_grad
-            # field, as ProcessGroupNCCL refuse to take a tensor with requires_grad
-            # to do inplace update! So we manually set it here
-            local_tensor.requires_grad_(tensor.requires_grad)
-            tensor = local_tensor
-        elif placement.is_replicate():
-            tensor = device_mesh.broadcast(tensor, mesh_dim=idx)
-        else:
-            raise RuntimeError("Not supported!")
-
-    return DTensor(
-        tensor, device_mesh, placements, requires_grad=tensor.requires_grad
-    )
 
+    if device_mesh is None:
+        device_mesh = get_global_device_mesh()
 
-# pyre-fixme[3]: Return type must be annotated.
-def distribute_module(
-    mod: nn.Module,
-    device_mesh: Optional[DeviceMesh] = None,
-    spec: Optional[Sequence[Placement]] = None,
-):
-    """
-    this function coverts all module parameters
-    to distributed tensor parameters according to
-    the placements and device_mesh spcified.
-    TODO: add a more flexible tagging, i.e. convert
-    certain param to a certain spec, like a PlacementPlan
-    """
+    def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
+        # This function loop over the immediate module parameters and
+        # buffers, replicate all non DTensor params/buffers to DTensor
+        # parameters/buffers, if they have not been partitioned in the
+        # partition_fn, we can't easily use `module._apply` here
+        # because we don't know what happened inside partition_fn as
+        # user could do anything, i.e. install hooks, and we want to
+        # preserve those.
+        full_replicate = [Replicate()] * mesh.ndim
+        for key, param in m._parameters.items():
+            if param is not None and not isinstance(param, DTensor):
+                submod.register_parameter(
```

Comment:
it surprises me that `submod` is actually in scope. Can someone explain this to me because I thought this should be `m` not `submod` but the code passed tests and ran without a problem. @wanchaol 

same question on line 62.

Reply:
Yeah I think it should be `m` actually, the reason that this get passed in tests is because python eval values dynamically, when this function get called, `submod` already in the scope and the `submod` is actually equivalent to `m`

Reply:
Yes, `submod` somehow becomes in scope which I still don't get how. I tried to write the same code in Python but couldn't get it run. Never mind, not a big deal. Can I create a PR to rename `submod` to `m` on line 54 and 62?